### DISCUSSION TITLE
STM32H7 Auto Baud Rate Detection

### DISCUSSION THREAD

**Discussion by thestumbler** [modm-io/modm#1106] [General]
Going against the grain of the MODM UART system, I noticed the STM32H7's UARTs have auto baud rate detection. This isn't something I would normally use, this is exactly how the factory boot loader (which I am mimicking) works. Short of bypassing MODM completely, is there any support planned for enabling this feature for those MCUs which have it?

**Comment by salkinium** [comment]
> any support planned

No, but it sounds useful so I would want to merge a PR with it.

You could re-use the current API by passing a special `Autodetect` value as baudrate and then using an `if constexpr` switch inside the `initialize<>` function. So it would even be backward compatible and you could static assert on those MCUs/instances that do not support it (like the SPI in UART mode instances for example).

**Comment by thestumbler** [comment]
Do any other MCUs you've worked with have such a built-in feature? I haven't run into this before. 

In my oddball situation, it's nice that the feature supports the exact same technique that the factory boot loader uses -- receipt of a 0x7F to measure the baud rate, which isn't completely surprising since ST made both. There's a couple of other patterns as well that the USART auto baud detection circuitry can key on, user configurable via the registers. 

I'm not clear on how useful this is in real life, because in most cases you want a clearly defined baud rate. In the case of working on this boot loader, I've experienced a wide range of maximum baud rates that work reliably, and I've never been able to pin down exactly why. And with a boot loader specifically, I can see why automatic vs manual baud rate could be helpful -- on the one hand, you can't change it manually if the MCU is erased, but on the other hand, you're potentially sending many hundreds of kilobytes of data back and forth and you'd want to use as fast a speed as possible.

**Comment by thestumbler** [comment]
Somewhat related to this, how do you set baud rate at runtime? I thought I saw a discussion of that in the documentation before, but I can't find it now. I will be doing auto baud later, but for now I want the ability to specify it.

**Reply by salkinium** [reply]
Baudrate prescaler computation is compile time only, so you'd use a fixed switch case of all the baudrates you want to support, which compiles down to a prescaler look-up table essentially. This is usually fine for a few baudrates, however, if you want to support *any* baudrate, you need to adapt the `Uart::initialize` function to make it work at runtime.

**Reply by thestumbler** [reply]
Any examples of this lookup table technique in the examples folder? I think I grasp what you're saying.

**Reply by salkinium** [reply]
No, but [my old blog post has an example here](https://blog.salkinium.com/computing-baudrates-at-compile-time/#usage).
